ogre2/src/Ogre2WideAngleCamera.cc

```diff
@@ -0,0 +1,678 @@
+  }
+  const uint32_t currVisibilityMask = this->VisibilityMask();
+  for (uint32_t i = 0u; i < kWideAngleNumCubemapFaces; ++i)
+  {
+    this->dataPtr->cubePassSceneDef[i]->mVisibilityMask = currVisibilityMask;
+  }
+
+  this->scene->StartRendering(this->dataPtr->ogreCamera);
+
+  Ogre::vector<Ogre::TextureGpu *>::type swappedTargets;
+
+  this->dataPtr->ogreCompositorWorkspace->setEnabled(true);
+
+  this->dataPtr->ogreCompositorWorkspace->_validateFinalTarget();
+  this->dataPtr->ogreCompositorWorkspace->_beginUpdate(false);
+  this->dataPtr->ogreCompositorWorkspace->_update();
+  this->dataPtr->ogreCompositorWorkspace->_endUpdate(false);
+
+  swappedTargets.clear();
+  this->dataPtr->ogreCompositorWorkspace->_swapFinalTarget(swappedTargets);
+
+  this->dataPtr->ogreCompositorWorkspace->setEnabled(false);
+
+  this->scene->FlushGpuCommandsAndStartNewFrame(6u, false);
+  {
+    auto engine = Ogre2RenderEngine::Instance();
+    auto ogreRoot = engine->OgreRoot();
+    ogreRoot->getRenderSystem()->endGpuDebuggerFrameCapture(nullptr);
+  }
+}
+
+//////////////////////////////////////////////////
+math::Vector3d Ogre2WideAngleCamera::Project3d(const math::Vector3d &_pt) const
+{
+  using namespace Ogre;
+
+  // Matches CompositorPass::CubemapRotations from OgreCompositorPass.cpp
+  const Quaternion kCubemapRotations[6] = {
+    Quaternion(Radian(-1.570796f), Vector3(0, 1, 0)),  // +X
+    Quaternion(Radian(1.570796f), Vector3(0, 1, 0)),   // -X
+    Quaternion(Radian(1.570796f), Vector3(1, 0, 0)),   // +Y
+    Quaternion(Radian(-1.570796f), Vector3(1, 0, 0)),  // -Y
+    Quaternion(1, 0, 0, 0),                            // +Z
+    Quaternion(Radian(3.1415927f), Vector3(0, 1, 0))   // -Z
+  };
+
+  this->dataPtr->ogreCamera->setAspectRatio(1.0f);
+  this->dataPtr->ogreCamera->setFOVy(Ogre::Degree(90));
+  const Quaternion oldCameraOrientation(
+    this->dataPtr->ogreCamera->getOrientation());
+
+  const bool isReverseDepth =
+    Root::getSingleton().getRenderSystem()->isReverseDepth();
+
+  // project world point to camera clip space.
+  const Matrix4 projMatrix = this->dataPtr->ogreCamera->getProjectionMatrix();
+  // const Vector4 basePos = viewProj * Vector4(Ogre2Conversions::Convert(_pt));
+  const Vector3 basePos = Ogre2Conversions::Convert(_pt);
+
+  // project onto cubemap face then onto
+  gz::math::Vector3d screenPos;
+  // loop through all env cameras can find the one that sees the 3d world point
+  for (unsigned int i = 0u; i < kWideAngleNumCubemapFaces; ++i)
+  {
+    // project world point to camera clip space for each face.
+    this->dataPtr->ogreCamera->setOrientation(oldCameraOrientation *
+                                              kCubemapRotations[i]);
+    const Matrix4 viewProj =
+      projMatrix * this->dataPtr->ogreCamera->getViewMatrix();
+    Vector4 pos = viewProj * Vector4(basePos);
+    pos.x /= pos.w;
+    pos.y /= pos.w;
+    pos.z /= pos.w;
```

Comment:
The ogre1 version happens to work because for most projection parameters; the range [near_plane; near_plane * 2.0] maps to the NDC range [-1; 0] while the range [near_plane * 2.0; far_plane] maps to NDC range [0; 1].

Thus doing `pos.z > 0` works well with the bug being unnoticed as long as the near plane is a small value.

Reply:
Ohh cool, Thanks for the explanation.

Reply:
I've updated the code to check for pos.z being in range [-1; 1]